### PR TITLE
Add an endpoint for creating Machine Deployments

### LINKED ISSUE #2212: Create an endpoint for creating MachineDeployments
**User Story**
In order to introduce MachineDeployment support, we're going to need an API endpoint for the UI to add them.

The new endpoint shall be POST `/api/v1/projects/{project_id}/dc/{dc}/clusters/{cluster_id}/nodesets` analogous to the existing POST `/api/v1/projects/{project_id}/dc/{dc}/clusters/{cluster_id}/nodes` endpoint. Instead of a `Machine` CRD, it will create a `MachineSet`.

**Acceptance criteria**
 * A new `NodeSet` API structure exists to use for the new API endpoint.
 * A  POST `/api/v1/projects/{project_id}/dc/{dc}/clusters/{cluster_id}/nodesets` endpoint exists, and sending a `NodeSet` object to to it results in a `MachineDeployment`.
 * There is a unit test for the functionality

### PR DESCRIPTION
**Pull request by maciaszczykm**
**What this PR does / why we need it**: It adds an endpoint for creating MachineDeployments.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: Fixes https://github.com/kubermatic/kubermatic/issues/2212

**Special notes for your reviewer**: I have tested it with slight changes inside the dashboard: https://github.com/kubermatic/dashboard-v2/compare/create-md?expand=1#diff-656f0718f18ffb3e2b0413c67636627c

``` sh
maciaszczykm @ ~ $ k get machinedeployment -n kube-system
NAME                                            AGE
machinedeployment-kubermatic-pml76fngcb-2jpl4   1h
machinedeployment-kubermatic-pml76fngcb-kjbcx   1h
machinedeployment-kubermatic-pml76fngcb-wfzng   55m

maciaszczykm @ ~ $ k get machineset -n kube-system
NAME                                                       AGE
machinedeployment-kubermatic-pml76fngcb-2jpl4-694ffdd76f   1h
machinedeployment-kubermatic-pml76fngcb-kjbcx-694ffdd76f   1h
machinedeployment-kubermatic-pml76fngcb-wfzng-694ffdd76f   55m

maciaszczykm @ ~ $ k get machine -n kube-system
NAME                                                             AGE
machinedeployment-kubermatic-pml76fngcb-2jpl4-694ffdd76f-6wkr7   1h
machinedeployment-kubermatic-pml76fngcb-kjbcx-694ffdd76f-44sqx   1h
machinedeployment-kubermatic-pml76fngcb-wfzng-694ffdd76f-bz9xs   55m
```

Not sure where we should set the strategy and labels, but it needs to be done as nil pointer dereference errors occur. What do you think about setting it to some default value if a request does not have them specified?

Release notes will be added in last PR.

**Documentation**: n/a

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: api/pkg/api/v1/types.go
Comment (alvaroaleman) [inline api/pkg/api/v1/types.go]: Do we want to use a pointer here? I know the underlying type does, but IMHO this should be a required argument

File: api/pkg/api/v1/types.go
Comment (alvaroaleman) [inline api/pkg/api/v1/types.go]: Right now this is required as we have no defaulting yet, but IMHO this shouldn't be required once we have that defaulting

Same is true for all the other fields below this comment, most users won't care I think

File: api/pkg/handler/node_v1.go
Comment (alvaroaleman) [inline api/pkg/handler/node_v1.go]: Shouldn't we check if that name is already taken?

File: api/pkg/api/v1/types.go
Comment (alvaroaleman) [inline api/pkg/api/v1/types.go]: This is certainly a nitpicky comment, but what about naming it `NodeGroup` because the fact that `NodeSet` has `Set` in it makes me think it would correspond to a `MachineSet`. WDYT?

**Review by alvaroaleman — COMMENTED** [review]
Some nits, but overall awesome work!

File: api/pkg/api/v1/types.go
Comment (maciaszczykm) [reply api/pkg/api/v1/types.go]: +1 from me. @kdomanski is it okay for you?

File: api/pkg/handler/node_v1.go
Comment (maciaszczykm) [reply api/pkg/handler/node_v1.go]: Yes, I will do it.

File: api/pkg/api/v1/types.go
Comment (maciaszczykm) [reply api/pkg/api/v1/types.go]: Okay, so I will implement the defaulting on API side.

File: api/pkg/api/v1/types.go
Comment (maciaszczykm) [reply api/pkg/api/v1/types.go]: Good point 👍  I will change it.

File: api/pkg/api/v1/types.go
Comment (alvaroaleman) [reply api/pkg/api/v1/types.go]: Thats sort of redundant once we have that webhook, maybe keep it required for now and change that once the webhook is part of kubermatic?

File: api/pkg/handler/node_v1.go
Comment (mrIncompetent) [reply api/pkg/handler/node_v1.go]: What for?
If its already taken, the apiserver will tell us, and we can pass through that error to the user.
There's no benefit it adding one additional API call before, just to make sure.

We can check the error with `IsAlreadyExists`.
So we can return the same error message as if we checked before.

File: api/pkg/api/v1/types.go
Comment (mrIncompetent) [reply api/pkg/api/v1/types.go]: Should we rather go with `MachineDeployment` in the API ?

I don't see any value in changing the name here. Instead it will become complicated in case we want to handle `MachineSet's` later on.

File: api/pkg/handler/node_v1.go
Comment (alvaroaleman) [reply api/pkg/handler/node_v1.go]: @mrIncompetent So you want endusers to see an error `Can not create NodeGroup: MachineDeployment machinedeployment-$rand-$rand already exists` and afterwards have them do the exact same thing again, this time without an error?

Comment (mrIncompetent) [reply api/pkg/handler/node_v1.go]: Ah right, you're referring to the random name generation! Missed that.

But yes, i would avoid the extra complexity for that. Using 5 random chars here leads to ~14 million possibilities. But not a strong opinion

File: api/pkg/api/v1/types.go
Comment (maciaszczykm) [reply api/pkg/api/v1/types.go]: @mrIncompetent Do you mean naming it `MachineDeployment` or directly returning `MachineDeployment`?

Good thing with keeping public structure might be possibility to filter/alter some fields and consistency (similar to `Node`).  Bad side of it will be the complexity when we would like to use machine-controller's `MachineSets`.

File: api/pkg/api/v1/types.go
Comment (maciaszczykm) [reply api/pkg/api/v1/types.go]: I have already done it as it is a minor change and I think having it can help. If you don't agree then I can revert.

File: api/pkg/api/v1/types.go
Comment (maciaszczykm) [reply api/pkg/api/v1/types.go]: Done.

File: api/pkg/handler/node_v1.go
Comment (maciaszczykm) [reply api/pkg/handler/node_v1.go]: 14 milion is a lot. Not sure about it anymore. Same thing is done for nodes at the moment.

**Comment by maciaszczykm** [comment]
Thank you for review @alvaroaleman and @mrIncompetent. Can you take another look? I will prepare tests tomorrow.

File: api/pkg/api/v1/types.go
Comment (alvaroaleman) [reply api/pkg/api/v1/types.go]: Okay, lets just keep it then :)

File: api/pkg/handler/node_v1.go
Comment (alvaroaleman) [reply api/pkg/handler/node_v1.go]: I'd just like to avoid the chance of random failures. AFAIK it should also be possible to use `metadata.GenerateName` which basically has this exact purpose: Generate a random name that is not used yet.

Node names are currently the same as machine names, so as long as we have no duplicate machines nodes will be unique.

File: api/pkg/resources/machine/machinedeployment.go
Comment (alvaroaleman) [inline api/pkg/resources/machine/machinedeployment.go]: Can we add a `TODO: Remove once we have webhook` here thought?

**Comment by alvaroaleman** [comment]
Modulo the unique name topic and the rename from `NodeSet` to something that doesn't have `Set` in it, LGTM :+1:

**Comment by maciaszczykm** [comment]
@mrIncompetent @alvaroaleman What about `NodeDeployment`? It would show that it is different from `MachineDeployment` but tell that it is equivalent of it.

File: api/pkg/handler/node_v1.go
Comment (mrIncompetent) [reply api/pkg/handler/node_v1.go]: Great idea! `metadata.GenerateName` would be perfect!

File: api/pkg/api/v1/types.go
Comment (mrIncompetent) [reply api/pkg/api/v1/types.go]: My idea was just about naming, not about directly passing through the `MachineDeployment`.
Passing through the whole object is something which caused some trouble over the past.

Comment (maciaszczykm) [reply api/pkg/api/v1/types.go]: Ah, okay. So I will adjust the naming.

File: api/pkg/resources/machine/machinedeployment.go
Comment (maciaszczykm) [reply api/pkg/resources/machine/machinedeployment.go]: Done.

File: api/pkg/api/v1/types.go
Comment (maciaszczykm) [reply api/pkg/api/v1/types.go]: Done.

File: api/pkg/handler/node_v1.go
Comment (maciaszczykm) [reply api/pkg/handler/node_v1.go]: Yeah, seems like a perfect fit. I have adjusted generating names of machine deployments but also for machines. Thanks again for an idea!

Sample name that is generated: `machinedeployment-kubermatic-x766wbbq4q-8xsfg-f965b978d-wxz78`, where `x766wbbq4q` part is the cluster ID.

**Review by mrIncompetent — DISMISSED** [review]

**Comment by maciaszczykm** [comment]
PTAL

**Review by mrIncompetent — APPROVED** [review]
Really, really cool!

**Comment by maciaszczykm** [comment]
Thank you. I cannot merge without @p0lyn0mial review :)

**Comment by mrIncompetent** [comment]
> Thank you. I cannot merge without @p0lyn0mial review :) 

I can :)
In case you're fine with skipping his review, ping me & i'll merge it

**Comment by maciaszczykm** [comment]
@mrIncompetent It was his recommendation that you will review it during his absence. He will be back after 2 weeks, so I think it would be good to merge this and continue working on next endpoints :)

**Pull request merged**
